src/NServiceBus.AzureStoragePersistence/CloudTableExtensions.cs

```diff
@@ -16,9 +18,18 @@ public static class CloudTableExtensions
             {
                 var seg = await table.ExecuteQuerySegmentedAsync(query, token, ct).ConfigureAwait(false);
                 token = seg.ContinuationToken;
-                items.AddRange(seg);
 
-            } while (token != null && !ct.IsCancellationRequested);
+                if (items.Count + seg.Count() > take)
```

Comment:
Excellent catch! I didn't look deep enough at the `TableQuerySegment<T>` object. It actually has a `Results` property which is a `List<T>` which is much better than my original attempt.